### PR TITLE
P3-231 Republish overwriting the original post

### PR DESCRIPTION
**Pull request by afercia**
## Context

* We want to empower users to publish the edits made on a copy onto the original post.

## Summary

This PR can be summarized in the following changelog entry:

* Implements overwriting the original post operation for the Rewrite and Republish feature.

## Notes
- When saving the copy as a "Draft" or "Pending review" or "Private", overwriting the original fails. This should be addressed separately.
- The "sticky" setting doesn't get copied, this needs a decision and should be addressed separately.
- Capabilities are not handled in this PR.
- As agreed, not all the new methods have a corresponding unit test. This will be addressed later.

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

**Preliminary steps:**
- Switch to this branch and build with `composer install && yarn && grunt build:js`.
- Install and activate Advanced Custom Fields.
- Create a simple custom fields group which contains, for example, an input field and a checkbox (see screenshot below)

acf field group example

- Set the custom fields group to be visible for all posts by setting a rule like in the screenshot below:

Screenshot 2020-12-10 at 11 20 26

- Deactivate Advanced Custom Fields for now.

**1 - Test with Classic Editor and no other plugins activated:**
- Install and activate the Classic Editor plugin.
- Pick a published post
- Take note of the post ID
- Create a copy for Rewrite & Republish.
- You will be redirected to the copy you just created.
- Check at the top of the screen a warning notice is displayed with the text "You can now start rewriting your post in this duplicate of the original post. If you click "Republish", your changes will be merged into the original post and you’ll be redirected there."
-  Take note of the copy ID visible in the browser's address bar.
- Check in the database (in the `wp_posts` table) the post copy has a `post_status` that is ~`dp-rewrite-draft`~ `draft` (this part was changed after CR feedback).
- Check in the database the original post has a meta `_dp_has_rewrite_republish_copy` with a value that is the copy ID (in the `wp_postmeta` table, search for `post_id` equal to the original post ID to see that post's meta)
- In the "Screen Options" at the top of the screen, make all boxes visible.
- Edit all the post properties that are meant to be copied onto the original:
  - post elements like:
    - title
	- content
	- excerpt
	- author
	- published date (set it to a past date! scheduled post with a future date will be addressed separately)
	- allow comments
	- allow trackbacks and pingbacks
	- optionally, change the post slug
  - taxonomies:
    - add or change existing categories
	- add or change existing tags
	- optionally: change post format (you will need a theme that supports post formats e.g. Twenty Thirteen)
  - post meta:
    - featured image 
	- template (you will need a theme that supports templates, e.g. Twenty Twenty)
	- add a custom field from the built-in WordPress "Custom Fields" meta box
- Publish.
- You will be redirected to the original post.
- Check at the top of the screen a success notice is displayed with the text "Your original post has been replaced with the rewritten post. You are now viewing the (rewritten) original post."
- Check that the original post has been overwritten with all the edits you made on the copy except the post slug: it should always be the slug of the original post.
- Check in the database the post copy has been deleted.
- Check in the database the original post has no `_dp_has_rewrite_republish_copy` meta any longer.

**2 - Test with Classic Editor and other plugins activated to verify the copying of custom meta:**
- Activate Advanced Custom Fields.
- Activate Yoast SEO.
- Pick a published post, create a copy for Rewrite & Republish, and repeat the steps above.
- This time, edit also the Yoast SEO and the ACF custom fields group values, for example:
  - focus keyphrase
  - seo title 
  - seo description 
  - cornerstone content 
  - some of the settings under "Advanced"
  - FB and Twitter image, title, description, under the "Social" tabs
  - the input field and checkbox in the ACF custom fields group
- Publish.
- Check everything works as expected and all the edits _including the ones from Yoast SEO and ACF_ are copied onto the original.

**3 - Test with Gutenberg and no other plugins activated:**
- Deactivate all plugins but Duplicate Post.
- Pick a published post and create a copy for Rewrite & Republish.
- Basically, repeat the steps for the case 1) 
- Check everything works as expected and all the edits are copied onto the original.

**4 - Test with Gutenberg and other plugins activated to verify the copying of custom meta:**
- Basically, repeat the steps for the case 2) 
- Check everything works as expected and all the edits _including the ones from Yoast SEO and ACF_ are copied onto the original.

Optionally, test with other post types e.g.: pages or books and movies (provided by the Yoast Test Helper plugin)

As a last step, verify the unit tests pass.

### Test instructions for QA when the code is in the RC

* [ ] QA should use the same steps as above.

QA can test this PR by following these steps:

* **Not applicable:** a PR with extensive test instructions for the whole Rewrite and Republish feature (or for the whole plugin, to check for regressions etc.) will follow.

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes

* [ ] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Documentation

* [ ] I have written documentation for this change.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [x] I have added unittests to verify the code works as intended

Fixes [P3-231]

### CODE REVIEW

**Comment by afercia** [comment]
Note: after https://github.com/Yoast/duplicate-post/pull/63 this PR needs to be adjusted to fix merge conflicts.

File: src/class-post-republisher.php [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+	/**
+	 * Republishes the original post with the passed post, when using the Block Editor.
+	 *
+	 * @param \WP_Post $post_data The copy's post object.
+	 *
+	 * @return void
+	 */
+	public function republish_after_rest_api_request( $post_data ) {
+		$this->republish_request( $post_data->ID, $post_data );
```
Comment (enricobattocchi) [inline src/class-post-republisher.php:181]: See above, a simple  `$post` object should be enough

File: src/class-post-republisher.php [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+	 * @param \WP_Post $post_data The copy's post object.
+	 *
+	 * @return void
+	 */
+	public function republish_after_post_request( $post_id, $post_data ) {
+		if ( defined( 'REST_REQUEST' ) && REST_REQUEST ) {
+			return;
+		}
+
+		$this->republish_request( $post_id, $post_data );
```
Comment (enricobattocchi) [inline src/class-post-republisher.php:200]: See above, a simple  `$post` object should be enough
The same for most (all?) occurrences of `$post_id, $post_data` as parameters later in the code

File: src/class-post-republisher.php [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+	 * @param int      $post_id   The copy's post ID.
+	 * @param \WP_Post $post_data The copy's post object.
+	 *
+	 * @return void
+	 */
+	protected function republish_post_meta( $post_id, $post_data ) {
+		$original_post_id = Utils::get_original_post_id( $post_id );
+
+		$copy_meta_options = [
+			'meta_excludelist' => [
```
Comment (enricobattocchi) [inline src/class-post-republisher.php:263]: I'm thinking about a single point of storage of these recurring field names to be filtered out by default...
Not for this issue, but as a later refinement

File: src/class-utils.php [hunk 4]
```diff
@@ -147,4 +147,15 @@ public static function get_edit_or_view_link( $post ) {
...
 	}
+
+	/**
+	 * Gets the ID of the original post intended to be rewritten with the copy for Rewrite & Republish.
+	 *
+	 * @param int $post_id The copy post ID.
+	 *
+	 * @return int The original post id of a copy for Rewrite & Republish.
+	 */
+	public static function get_original_post_id( $post_id ) {
```
Comment (enricobattocchi) [inline src/class-utils.php:158]: Why not use `Utils::get_original()` and then get the `ID` from the resulting post object?

File: src/ui/class-block-editor.php [hunk 5]
```diff
@@ -93,10 +99,12 @@ public function enqueue_block_editor_scripts() {
 
 		\wp_localize_script(
 			'duplicate_post_edit_script',
-			'duplicatePostLinks',
+			'duplicatePost',
 			[
 				'new_draft_link'             => $this->get_new_draft_permalink(),
 				'rewrite_and_republish_link' => $this->get_rewrite_republish_permalink(),
+				'rewriting'                  => \get_post_meta( $post->ID, '_dp_is_rewrite_republish_copy', true ) ? 1 : 0,
```
Comment (enricobattocchi) [inline src/ui/class-block-editor.php:106]: Not sure why the `is_rewrite_and_republish_copy()` method from Permissions_Helper is not used...

File: src/ui/class-block-editor.php [hunk 6]
```diff
@@ -132,4 +140,32 @@ public function get_rewrite_republish_permalink() {
...
+	 * @return string The URL. Empty if the copy post doesn't have an original.
+	 */
+	public function get_original_post_edit_url() {
+		$post = \get_post();
+
+		if ( \is_null( $post ) ) {
+			return '';
+		}
+
+		$original_post_id = Utils::get_original_post_id( $post->ID );
```
Comment (enricobattocchi) [inline src/ui/class-block-editor.php:156]: See above about `Utils::get_original_post_id` when we already have `Utils::get_original`

**Review by enricobattocchi — COMMENTED** [review]
CR :construction: 
Some questions and a bit of polishing suggested.

File: src/class-post-duplicator.php [hunk 7]
```diff
@@ -147,14 +147,15 @@ public function set_modified( $data ) {
...
-			'use_filters'     => false,
+			'copy_title'        => true,
+			'copy_date'         => true,
+			'copy_name'         => true,
+			'copy_content'      => true,
+			'copy_excerpt'      => true,
+			'copy_author'       => true,
+			'copy_menu_order'   => true,
+			'use_filters'       => false,
+			'force_post_status' => 'dp-rewrite-draft',
```
[the remark this answers is not in the document]
Comment (afercia) [reply src/class-post-duplicator.php:158]: Will remove 👍

File: src/class-post-republisher.php [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+
+	/**
+	 * Executes the republish request.
+	 *
+	 * @param int      $post_id   The copy's post ID.
+	 * @param \WP_Post $post_data The copy's post object.
+	 *
+	 * @return void
+	 */
+	public function republish_request( $post_id, $post_data ) {
```
[the remark this answers is not in the document]
Comment (afercia) [reply src/class-post-republisher.php:146]: Remnant of previous implementation. Will double check.

File: src/ui/class-block-editor.php (hunk 5 above)
Comment (afercia) [reply src/ui/class-block-editor.php:106]: Because it didn't exist when this PR started 🙂

File: src/class-utils.php (hunk 4 above)
Comment (afercia) [reply src/class-utils.php:158]: `Utils::get_original()` returns  `WP_Post|null` so before getting the ID I'd need to check for `null` for all the three times it's used in this class. Seems to me that would be unnecessarily repeated code. Instead, `Utils::get_original_post_id()` already does what I need. It always returns an `int`. In fact, if `get_post_meta()` fails, `false` is casted to a zero integer .

**Comment by enricobattocchi** [comment]
CR 2 OK :+1:

**Comment by afercia** [comment]
Updated the test instructions because the custom post status `dp-rewrite-draft` was removed after CR feedback.

**Comment by jcomack** [comment]
Acceptance done 🚧 

Overall, most of the steps are correct / works, however there are a few things that stand out:

* The step `Take note of the original post ID and the copy ID visible in the browser's address bar.` no longer applies. This has already been discussed with @afercia and should be removed.
* Changing the slug and publishing the post, does not work. This applies to both the Classic and Block Editor. Other fields seem to work fine.
* The database seems to end up containing some 'old' or seemingly duplicate data, after you've published. See screenshots below.

Post has multiple revisions, generated rapidly after one another and also have the same post name (happened with the Block Editor)
_MySQL_5_7_26__basic_wordpress_test_wordpress_wp_posts

Old post revisions and republished post revisions are still present in the posts table (happened with the Classic Editor)
_MySQL_5_7_26__basic_wordpress_test_wordpress_wp_posts

**Comment by afercia** [comment]
- Clarified test instructions about `Take note of the original post ID ...`.
- Clarified test instructions about the post slug: actually, this is the only post property that should not change. When the copy overwrites the original, we make sure and enforce via code the original post keeps its original slug.
- Revisions: this is known and it's just how WordPress works. When publishing a copy to overwrite the original:
  - we first save the copy so that the copy post is updated with all the edits: this also creates a revision of the copy
  - we overwrite the original with the copy by using `wp_update_post()` (which internally uses `wp_insert_post()`): this creates a new revision of the original.

**Comment by jcomack** [comment]
After the clarifications and a second round of testing, everything seems clear now. Merging!

**Pull request merged**